src/main/java/com/google/sps/recommendation/ProcessNearestNeighbors.java

```diff
@@ -0,0 +1,119 @@
+package com.google.sps.recommendation;
+
+import org.deeplearning4j.models.embeddings.inmemory.InMemoryLookupTable;
+import org.deeplearning4j.models.embeddings.wordvectors.WordVectors;
+import org.deeplearning4j.models.embeddings.loader.WordVectorSerializer;
+import org.deeplearning4j.models.word2vec.wordstore.VocabCache;
+
+import org.nd4j.linalg.api.buffer.DataType;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.stream.IntStream;
+import com.google.sps.tool.ArrayUtil;
+import org.mapdb.*;
+
+/* This function of this file is to ONLY pre-process the nearest neigbors of 300,000 words.
+ * It will create/update a binary .db file. Time results and optimal batch size
+ * are based on running this file on a Google Pixelbook. Time result may change for
+ * different hardwares.
+ *
+ * Processing Time:
+ * On a Google Pixelbook, a batch size of 500 words results in the fastest runtime for
+ * computing the nearest neighbors, while avoiding OOMs. We may see some speedup by
+ * migrating from N4dj matrix operations to a numerical computing library like
+ * numpy.
+ *
+ * Time example:
+ * Each batch of 500 words takes about 60,000 - 80,000 ms.
+ * Processing 100 batches can take from 1 hour 30 minutes - 2 hours.
+ *
+ * Some underlying reasons for this slow time performance:
+ * 1) N4dj's behind the hood operations. For example, to slice a large matrix by a range of indices,
```

Comment:
can you link to the resource (stack overflow) that documents this?

Reply:
done